8301997: Move method resolution information out of the cpCache

matias9927 opened the pull request:
The current structure used to store the resolution information for methods, ConstantPoolCacheEntry, is difficult to interpret due to its ambigious fields f1 and f2. This structure previously held information for fields, methods, and invokedynamic calls which were all encoded into f1 and f2. Currently this structure only handles method entries, but it remains obtuse and inconsistent with recent changes. 

This enhancement introduces a new data structure that stores the necessary resolution data in an intuitive an extensible manner. These resolved entries are stored in an array inside the constant pool cache in a very similar manner to invokedynamic entries in JDK-8301995.

Instances of ConstantPoolCache entry related to field resolution have been replaced with the new ResolvedMethodEntry, and ConstantPoolCacheEntry has been removed entirely. The class ResolvedMethodEntry holds resolution information for all types of invoke calls besides invokedynamic, and thus has fields that may be unused depending on the invoke code. 

To streamline the review, please consider these major areas that have been changed:
1. ResolvedMethodEntry class
2. Rewriter for initialization of the structure
3. cpCache for resolution
4. InterpreterRuntime, linkResolver, and templateTable
5. JVMCI
6. SA

Verified with tier 1-9 tests.

This change supports the following platforms: x86, aarch64, RISCV, PPC

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301997](https://bugs.openjdk.org/browse/JDK-8301997): Move method resolution information out of the cpCache (**Sub-task** - P4)


### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)
 * [Andrew Dinn](https://openjdk.org/census#adinn) (@adinn - **Reviewer**) ⚠️ Review applies to [7a7aff03](https://git.openjdk.org/jdk/pull/15455/files/7a7aff0382cb78788442a8ed7ff87fb0c63d89ab)
 * [Frederic Parain](https://openjdk.org/census#fparain) (@fparain - **Reviewer**) ⚠️ Review applies to [6950709c](https://git.openjdk.org/jdk/pull/15455/files/6950709c2fb12d9b0648e98048457e2306cc47a8)


### Contributors
 * Gui Cao `<gcao@openjdk.org>`
 * Fei Yang `<fyang@openjdk.org>`
 * Martin Doerr `<mdoerr@openjdk.org>`
 * Amit Kumar `<amitkumar@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15455/head:pull/15455` \
`$ git checkout pull/15455`

Update a local copy of the PR: \
`$ git checkout pull/15455` \
`$ git pull https://git.openjdk.org/jdk.git pull/15455/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15455`

View PR using the GUI difftool: \
`$ git pr show -t 15455`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15455.diff">https://git.openjdk.org/jdk/pull/15455.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15455#issuecomment-1765040952)

adinn commented on src/hotspot/cpu/aarch64/interp_masm_aarch64.cpp:
```diff
@@ -1866,3 +1797,16 @@ void InterpreterMacroAssembler::load_field_entry(Register cache, Register index,
+void InterpreterMacroAssembler::load_method_entry(Register cache, Register index, int bcp_offset) {
+  // Get index out of bytecode pointer
+  get_cache_index_at_bcp(index, bcp_offset, sizeof(u2));
+  // Take shortcut if the size is a power of 2
```
Comment needs removing (size is not a power of 2).

adinn commented on src/hotspot/cpu/aarch64/templateTable_aarch64.cpp:
```diff
@@ -2257,78 +2258,86 @@ void TemplateTable::_return(TosState state)
 // volatile-loads.
 
-void TemplateTable::resolve_cache_and_index(int byte_no,
+void TemplateTable::resolve_cache_and_index_for_field(int byte_no,
```
This change looks more confusing than it actually needs because you deleted the definition for `resolve_cache_and_index` that preceded `resolve_cache_and_index_for_field` and then added `resolve_cache_and_index_for_method` after `resolve_cache_and_index_for_field`. As a result we see changes made to `resolve_cache_and_index` that re-introduce changes already made for `resolve_cache_and_index_for_field` followed by changes to `resolve_cache_and_index_for_field` that rework it to remove field-specific operations and replace them with method-specific operations.

That's going to make it a lot harder for maintainers and back-porters to work out what this change is really doing. If instead you place the definition for `resolve_cache_and_index_for_method` before the definition for `resolve_cache_and_index_for_field` then the diff will show `resolve_cache_and_index` being repurposed to cater for methods and the definition of `resolve_cache_and_index_for_field` should remain unchanged.

matias9927 replied:
Great point! I didn't notice how messy the diff ended up looking.

adinn commented on src/hotspot/cpu/aarch64/templateTable_aarch64.cpp:
```diff
@@ -2357,32 +2366,111 @@ void TemplateTable::load_resolved_field_entry(Register obj,
-                                              Register flags,
-                                              bool is_static = false) {
-  assert_different_registers(cache, index, flags, off);
+void TemplateTable::load_resolved_method_entry_common(Register cache,
```
When I saw the `common` suffix I thought at first that this might be used in addition to the variants with the other suffixes rather than as an alternative. Since this is 'common' to only to 2 out of the 5 different cases perhaps it could be named `load_resolved_method_entry_special_or_static`. That would make it clear that each invoke type has a corresponding `load_resolved_method_entry_<invoketype>` variant.

adinn replied:
Yes, it's a long name but it clarifies the intended purpose consistently with the names for the other cases. I think it is best to stick with it.

adinn commented on src/hotspot/cpu/aarch64/templateTable_aarch64.cpp:
```diff
@@ -2357,32 +2366,111 @@ void TemplateTable::load_resolved_field_entry(Register obj,
+  // setup registers
+  const Register index = r4;
+  assert_different_registers(method_or_table_index, flags);
+  assert_different_registers(method_or_table_index, cache, flags);
```
This assert renders the previous one redundant

adinn commented on src/hotspot/cpu/aarch64/templateTable_aarch64.cpp:
```diff
@@ -2357,32 +2366,111 @@ void TemplateTable::load_resolved_field_entry(Register obj,
+  resolve_cache_and_index_for_method(f1_byte, cache, index);
+  __ load_unsigned_byte(flags, Address(cache, in_bytes(ResolvedMethodEntry::flags_offset())));
+
+  // table_or_ref_index can either be an itable index or a resolved reference index depending on the bytecode
```
This comment seems to apply to `ref_index` in `load_resolved_method_entry_handle`? Does it need to move up there? Do we need a comment here stating somethgin similar for `method_or_table_index`?

matias9927 replied:
This looks like a leftover from a previous implementation. I don't believe this is true for either invokeinterface or invokehandle.

adinn commented on src/hotspot/cpu/aarch64/templateTable_aarch64.cpp:
```diff
@@ -3242,54 +3292,24 @@ void TemplateTable::fast_xaccess(TosState state)
-    __ bind(L_no_push);
-  }
+  // Load TOS state for later
+  __ load_unsigned_byte(rscratch2, Address(cache, in_bytes(ResolvedMethodEntry::type_offset())));
```
In the old code this came after the `if (load_receiver)` block just before rscratch2 was used to compute the return address. This was for a good reason. At the end of the `if` block there is a call to `verify_oop` which uses rscratch2 when VerifyOops is set. So, relocating this load of the TOS state will fail to compute the correct return address in that case.

adinn replied:
Oops, ignore that! `verify_oops` pushes rscratch2 on the stack before using it and restores it afterwards.

adinn commented on src/hotspot/cpu/aarch64/templateTable_aarch64.cpp:
```diff
@@ -3242,54 +3292,24 @@ void TemplateTable::fast_xaccess(TosState state)
   if (load_receiver) {
-    __ andw(recv, flags, ConstantPoolCacheEntry::parameter_size_mask);
+    __ load_unsigned_short(recv, Address(cache, in_bytes(ResolvedMethodEntry::num_parameters_offset())));
     // FIXME -- is this actually correct? looks like it should be 2
```
This 'FIXME' comment (all 5 lines) does not need to be kept. Likewise the trailing comment that says 'FIXME: uxtb here?'

adinn commented on src/hotspot/cpu/aarch64/templateTable_aarch64.cpp:
```diff
@@ -3242,54 +3292,24 @@ void TemplateTable::fast_xaccess(TosState state)
 
   // save 'interpreter return address'
-  __ save_bcp();
+  __ save_bcp(); // probably don't need this
```
I'm not sure why you added this comment but we do need this call to `save_bcp()`. It records in the interpreter frame the 'return address' needed for re-entry into the interpreter. We cannot rely on the value cached in register rbcp remaining constant across the invoke that is about to happen.

adinn commented on src/hotspot/cpu/x86/templateTable_x86.cpp:
```diff
@@ -2854,37 +2835,111 @@ void TemplateTable::load_invokedynamic_entry(Register method) {
 }
 
-void TemplateTable::load_invoke_cp_cache_entry(int byte_no,
+void TemplateTable::load_resolved_method_entry_common(Register cache,
```
Same comment applies regarding the `common` suffix as for aarch64 code.

adinn commented on src/hotspot/cpu/x86/templateTable_x86.cpp:
```diff
@@ -2854,37 +2835,111 @@ void TemplateTable::load_invokedynamic_entry(Register method) {
+  resolve_cache_and_index_for_method(f1_byte, cache, index);
+  __ load_unsigned_byte(flags, Address(cache, in_bytes(ResolvedMethodEntry::flags_offset())));
+
+  // table_or_ref_index can either be an itable index or a resolved reference index depending on the bytecode
```
same comment applies as for aarch64 code

adinn commented on src/hotspot/cpu/x86/templateTable_x86.cpp:
```diff
@@ -3987,9 +4015,9 @@ void TemplateTable::invokedynamic(int byte_no) {
   load_invokedynamic_entry(rbx_method);
   // rax: CallSite object (from cpool->resolved_references[f1])
-  // rbx: MH.linkToCallSite method (from f2)
+  // rbx: MH.linkToCallSite method
```
I'm not sure why you edited these comments in the x86 code but not in the aarch64 code. To avoid issues for maintainers it is better to make the changes in both rather than just one.

adinn commented on src/hotspot/cpu/x86/templateTable_x86.cpp:
```diff
@@ -3987,9 +4015,9 @@ void TemplateTable::invokedynamic(int byte_no) {
+  // rbx: MH.linkToCallSite method
 
-  // Note:  rax_callsite is already pushed by prepare_invoke
+  // Note:  rax_callsite is already pushed
```
As per previous change.

adinn commented on src/hotspot/share/ci/ciReplay.cpp:
```diff
@@ -433,12 +432,12 @@ class CompileReplay : public StackObj {
-        cp_cache_entry = cp->cache()->entry_at(cp->decode_cpcache_index(index));
-        cp_cache_entry->set_method_handle(cp, callInfo);
+        ResolvedMethodEntry* method_entry = cp->cache()->resolved_method_entry_at(index);
+        cp->cache()->set_method_handle(index, callInfo);
```
It looks a bit odd that you obtain a pointer to the `ResolvedMethodEntry` at `index` by calling `resolved_method_entry_at` and then pass `index` back into `set_method_handle` in order to update it with the call info. Obviously this is done so that the set routine can handle data races. Would it not be better to modify `set_method_handle` so that it handled the race and also returned the `ResolvedMethodEntry` at `index`.

Likewise you pass `index` back in the call to `appendix_if_resolved` below. Would it not be better to have this method accept a `ResolvedMethodEntry` pointer?
Likewise

matias9927 replied:
Right, currently these calls seem redundant since it reads the resolved method entry twice. If I understand correctly, you are suggesting that the methods look something like this?
`ResolvedMethodEntry set_method_handle(int index, const CallInfo &call_info)`
`oop appendix_if_resolved(ResolvedMethodEntry* method_entry)`

adinn replied:
Yes, removing the redundant lookup is the correct rationale. However, I suggested a better solution below in a comment attached to method `ConstantPoolCache::appendix_if_resolved(int index)`.

Keep the current method of `ConstantPoolCache` that takes an index but modify it to delegate the appendix check to a new method `ResolvedMethodEntry::appendix_if_resolved()`. You can make a direct call to the latter method when you have already looked up the `ResolvedMethodEntry`.

coleenp reviewed (changes requested):
I'm about halfway through.

adinn commented on src/hotspot/share/interpreter/bytecodeTracer.cpp:
```diff
@@ -529,7 +524,7 @@ void BytecodePrinter::print_attributes(int bci, outputStream* st) {
           assert(is_linked(), "invokehandle is only in rewritten methods");
           assert(cpcache_index >= 0, "must be");
-          print_cpcache_entry(cpcache_index, st);
+          print_field_or_method(cp_index, st);
```
I don't understand this code very well but it looks like this change means `print_field_or_method` gets called twice when we have an `invokehandle` bytecode, passing `cp_index` both times. Is that intended?

adinn replied:
This comment is still left to address, Matias.

adinn commented on src/hotspot/share/interpreter/linkResolver.cpp:
```diff
@@ -1699,12 +1699,11 @@ void LinkResolver::resolve_invokeinterface(CallInfo& result, Handle recv, const
-    methodHandle method(THREAD, cpce->f1_as_method());
-    Handle     appendix(THREAD, cpce->appendix_if_resolved(pool));
+    methodHandle method(THREAD, method_entry->method());
+    Handle     appendix(THREAD, pool->cache()->appendix_if_resolved(index));
```
This is another point where you already have the `ResolvedMethodEntry` and are still looking up the appendix info via the cache using an index.

adinn commented on src/hotspot/share/interpreter/zero/bytecodeInterpreter.cpp:
```diff
@@ -2306,43 +2307,43 @@ void BytecodeInterpreter::run(interpreterState istate) {
           CALL_VM(InterpreterRuntime::resolve_from_cache(THREAD, (Bytecodes::Code)opcode),
                   handle_exception);
-          cache = cp->entry_at(index);
+          entry = cp->resolved_method_entry_at(index);
```
Do you actually need to lookup the entry again? I'm not really sure why the old code needed to do so.

matias9927 replied:
This is probably unnecessary. I will try removing it and see what the test results look like.

adinn commented on src/hotspot/share/jvmci/jvmciCompilerToVM.cpp:
```diff
@@ -1680,7 +1682,7 @@ C2V_VMENTRY_0(jint, isResolvedInvokeHandleInPool, (JNIEnv* env, jobject, ARGUMEN
       vmassert(MethodHandles::is_signature_polymorphic_method(resolved_method()),"!");
       vmassert(!MethodHandles::is_signature_polymorphic_static(resolved_method->intrinsic_id()), "!");
-      vmassert(cp_cache_entry->appendix_if_resolved(cp) == nullptr, "!");
+      vmassert(cp->cache()->appendix_if_resolved(index) == nullptr, "!");
```
Another place where you have the `ResolvedMethodEntry` and are accessing via the cache with an index.

offamitkumar commented:
Hi @matias9927, 

would you please rebase this PR with the head stream.

adinn commented on src/hotspot/share/oops/cpCache.cpp:
```diff
@@ -940,30 +638,41 @@ oop ConstantPoolCache::set_dynamic_call(const CallInfo &call_info, int index) {
 
+oop ConstantPoolCache::appendix_if_resolved(int method_index) const {
+  ResolvedMethodEntry* method_entry = resolved_method_entry_at(method_index);
+  if (!method_entry->has_appendix())
```
If you move the rest of the code in this method into the a new method `ResolvedMethodEntry::appendix_if_resolved()` then you can call that method from here and also call it in places where you have already looked up the `ResolvedMethodEntry` but are still indirecting through the cache method using an `index`.

matias9927 replied:
ResolvedMethodEntry doesn't know about the constant pool or the cache, so that mean it will be unable to call `constant_pool()->resolved_reference_at(ref_index);`. Maybe an alternate solution would be to overload `appendix_if_resolved` so that it can take either `int method_index` or `ResolvedMethodEntry* method_entry` as arguments.

adinn replied:
Doh! Yes, of course. Overloading will allow the repeated lookup to be avoided.

adinn commented on src/hotspot/share/oops/constantPool.inline.hpp:
```diff
@@ -51,6 +52,22 @@ inline int ConstantPool::resolved_field_entries_length() const {
+
+inline oop ConstantPool::appendix_if_resolved(int method_index) const {
+  ResolvedMethodEntry* entry = cache()->resolved_method_entry_at(method_index);
+  if (!entry->has_appendix())
```
As per the method on `ConstantPoolCache` if you move the body of this into `ResolvedMethodEntry::appendix_if_resolved` you can call that direct from here.

adinn reviewed (changes requested):
Hi Matias. I have reviewed the aarch64- and x86-specific changes and also worked through the shared code in c1, c1, classfile, interpreter, zero, oops and prims and have made some suggestions for changes. No doubt Coleen will pick up on things I am quite likely to have missed in the shared code.

adinn commented:
Oh, and I forgot to say, nice work!

matias9927 commented on src/hotspot/share/interpreter/rewriter.cpp:
```diff
@@ -220,28 +225,23 @@ void Rewriter::rewrite_invokespecial(address bcp, int offset, bool reverse, bool
         // allow invokespecial as an alias, although it would be very odd:
-        (*opc) == (u1)Bytecodes::_invokespecial) {
-      assert(_pool->tag_at(cp_index).is_method(), "wrong index");
+        ((*opc) == (u1)Bytecodes::_invokespecial && _pool->tag_at(cp_index).is_method())) {
```
This looks like a leftover of an earlier attempt to resolve the issue with invokespecial. I believe this can be reverted

adinn commented on src/hotspot/share/oops/cpCache.hpp:
```diff
@@ -38,196 +38,90 @@
-  friend class ConstantPool;
-  friend class InterpreterRuntime;
+// A constant pool cache is a runtime data structure set aside to a constant pool. The cache
+// holds interpreter runtime information for all field access and invoke bytecodes. The cache
```
I think there is a bit more nuance needed here. Strictly, the cpCache *is* a cache since it retains values for re-use that would otherwise need to be recomputed. What it is not is a traditional key-value lookup store i.e. it does not conform to the common usage of 'cache' in computer science parlance but to a more broad use of the term in English. I agree that we could do with a better name.

Also, while you are correct to note that the runtime uses the cpCache during link resolution and that the JITs use the cpCache when generating compiled code, it is also true that the resulting JITted method code does not refer to the cache as part of its operation. So, as far as the execution engine of the JVM is concerned, the info in the cpCache is present to optimize interpreted execution and is incidental to compiled execution. That distinction explains why it has been referred to as an interpreter 'cache' or 'scratch pad' and I think this is worth noting, albeit at the risk of down-playing those other uses. This present discussion may well all that is needed to provide that note so, as you say, the comment should probably avoid tying the cpCache exclusively to the interpreter.

matias9927 commented on src/hotspot/share/oops/cpCache.cpp:
```diff
@@ -299,86 +287,67 @@ void ConstantPoolCacheEntry::set_direct_or_vtable_call(Bytecodes::Code invoke_co
-  // There are three words to update: flags, refs[f2], f1 (in that order).
-  // Writers must store all other values before f1.
-  // Readers must test f1 first for non-null before reading other fields.
+void ConstantPoolCache::set_method_handle_common(int method_index,
```
I was also thinking about this after seeing @adinn 's comments. Neither function is private or protected so I think it is safe to just move all of the code from common to `set_method_handle()`

zifeihan commented:
Hello, I'm preparing the riscv part of the support, but I noticed that the x86 backend is reporting errors.
```
zifeihan@plct-c8:~/jdk$ make test TEST="tier1" JTREG="TIMEOUT_FACTOR=16"
Building target 'test' in configuration 'linux-x86_64-server-release'
Running tests using JTREG control variable 'TIMEOUT_FACTOR=16'
Test selection 'tier1', will run:
* jtreg:test/hotspot/jtreg:tier1
* jtreg:test/jdk:tier1
* jtreg:test/langtools:tier1
* jtreg:test/jaxp:tier1
* jtreg:test/lib-test:tier1

Running test 'jtreg:test/hotspot/jtreg:tier1'
An exception has occurred in the compiler (22-internal). Please file a bug against the Java compiler via the Java bug reporting page (https://bugreport.java.com) after checking the Bug Database (https://bugs.java.com) for duplicates. Include your program, the following diagnostic, and the parameters passed to the Java compiler in your report. Thank you.
java.lang.invoke.WrongMethodTypeException: cannot convert MethodHandle(VarHandle,byte[],int)long to (VarHandle,byte[],int)int
	at java.base/java.lang.invoke.MethodHandle.asTypeUncached(MethodHandle.java:903)
	at java.base/java.lang.invoke.MethodHandle.asType(MethodHandle.java:870)
	at java.base/jdk.internal.util.ByteArray.getLong(ByteArray.java:188)
	at java.base/java.io.DataInputStream.readLong(DataInputStream.java:408)
	at jdk.compiler/com.sun.tools.javac.util.ByteBuffer.getLong(ByteBuffer.java:211)
	at jdk.compiler/com.sun.tools.javac.jvm.PoolReader.resolve(PoolReader.java:260)
	at jdk.compiler/com.sun.tools.javac.jvm.PoolReader$ImmutablePoolHelper.readIfNeeded(PoolReader.java:391)
	at jdk.compiler/com.sun.tools.javac.jvm.PoolReader.getConstant(PoolReader.java:206)
	at jdk.compiler/com.sun.tools.javac.jvm.ClassReader$3.read(ClassReader.java:855)
	at jdk.compiler/com.sun.tools.javac.jvm.ClassReader.readAttrs(ClassReader.java:1433)
	at jdk.compiler/com.sun.tools.javac.jvm.ClassReader.readMemberAttrs(ClassReader.java:1423)
	at jdk.compiler/com.sun.tools.javac.jvm.ClassReader.readField(ClassReader.java:2283)
	at jdk.compiler/com.sun.tools.javac.jvm.ClassReader.readClass(ClassReader.java:2645)
	at jdk.compiler/com.sun.tools.javac.jvm.ClassReader.readClassBuffer(ClassReader.java:2738)
	at jdk.compiler/com.sun.tools.javac.jvm.ClassReader.readClassFile(ClassReader.java:2762)
	at jdk.compiler/com.sun.tools.javac.code.ClassFinder.fillIn(ClassFinder.java:373)
	at jdk.compiler/com.sun.tools.javac.code.ClassFinder.complete(ClassFinder.java:302)
	at jdk.compiler/com.sun.tools.javac.code.Symtab$2.complete(Symtab.java:360)
	at jdk.compiler/com.sun.tools.javac.code.Symbol.complete(Symbol.java:682)
	at jdk.compiler/com.sun.tools.javac.code.Symbol$ClassSymbol.complete(Symbol.java:1418)
	at jdk.compiler/com.sun.tools.javac.code.Symbol$ClassSymbol.flags(Symbol.java:1334)
	at jdk.compiler/com.sun.tools.javac.code.Types$12.visitClassType(Types.java:2183)
	at jdk.compiler/com.sun.tools.javac.code.Types$12.visitClassType(Types.java:2159)
	at jdk.compiler/com.sun.tools.javac.code.Type$ClassType.accept(Type.java:1050)
	at jdk.compiler/com.sun.tools.javac.code.Types$DefaultTypeVisitor.visit(Types.java:4894)
	at jdk.compiler/com.sun.tools.javac.code.Types.asSuper(Types.java:2156)
	at jdk.compiler/com.sun.tools.javac.code.Types$12.visitClassType(Types.java:2179)
	at jdk.compiler/com.sun.tools.javac.code.Types$12.visitClassType(Types.java:2159)
	at jdk.compiler/com.sun.tools.javac.code.Type$ClassType.accept(Type.java:1050)
```

adinn commented on src/hotspot/share/oops/resolvedMethodEntry.hpp:
```diff
@@ -0,0 +1,208 @@
+  void remove_unshareable_info();
+
+  // Offsets
+  static ByteSize klass_offset()                     { return byte_offset_of(ResolvedMethodEntry, _entry_specific._interface_klass); }
```
Perhaps the getters for the entry specific union fields should include an assert to verify that the flags are consistent with the access being performed?

matias9927 commented:
> Hello, I'm preparing the riscv part of the support, but I noticed that the x86 backend is reporting errors.
> 
> ```
> zifeihan@plct-c8:~/jdk$ make test TEST="tier1" JTREG="TIMEOUT_FACTOR=16"
> Building target 'test' in configuration 'linux-x86_64-server-release'
> Running tests using JTREG control variable 'TIMEOUT_FACTOR=16'
> Test selection 'tier1', will run:
> * jtreg:test/hotspot/jtreg:tier1
> * jtreg:test/jdk:tier1
> * jtreg:test/langtools:tier1
> * jtreg:test/jaxp:tier1
> * jtreg:test/lib-test:tier1
> 
> Running test 'jtreg:test/hotspot/jtreg:tier1'
> An exception has occurred in the compiler (22-internal). Please file a bug against the Java compiler via the Java bug reporting page (https://bugreport.java.com) after checking the Bug Database (https://bugs.java.com) for duplicates. Include your program, the following diagnostic, and the parameters passed to the Java compiler in your report. Thank you.
> java.lang.invoke.WrongMethodTypeException: cannot convert MethodHandle(VarHandle,byte[],int)long to (VarHandle,byte[],int)int
> 	at java.base/java.lang.invoke.MethodHandle.asTypeUncached(MethodHandle.java:903)
> 	at java.base/java.lang.invoke.MethodHandle.asType(MethodHandle.java:870)
> 	at java.base/jdk.internal.util.ByteArray.getLong(ByteArray.java:188)
> 	at java.base/java.io.DataInputStream.readLong(DataInputStream.java:408)
> 	at jdk.compiler/com.sun.tools.javac.util.ByteBuffer.getLong(ByteBuffer.java:211)
> 	at jdk.compiler/com.sun.tools.javac.jvm.PoolReader.resolve(PoolReader.java:260)
> 	at jdk.compiler/com.sun.tools.javac.jvm.PoolReader$ImmutablePoolHelper.readIfNeeded(PoolReader.java:391)
> 	at jdk.compiler/com.sun.tools.javac.jvm.PoolReader.getConstant(PoolReader.java:206)
> 	at jdk.compiler/com.sun.tools.javac.jvm.ClassReader$3.read(ClassReader.java:855)
> 	at jdk.compiler/com.sun.tools.javac.jvm.ClassReader.readAttrs(ClassReader.java:1433)
> 	at jdk.compiler/com.sun.tools.javac.jvm.ClassReader.readMemberAttrs(ClassReader.java:1423)
> 	at jdk.compiler/com.sun.tools.javac.jvm.ClassReader.readField(ClassReader.java:2283)
> 	at jdk.compiler/com.sun.tools.javac.jvm.ClassReader.readClass(ClassReader.java:2645)
> 	at jdk.compiler/com.sun.tools.javac.jvm.ClassReader.readClassBuffer(ClassReader.java:2738)
> 	at jdk.compiler/com.sun.tools.javac.jvm.ClassReader.readClassFile(ClassReader.java:2762)
> 	at jdk.compiler/com.sun.tools.javac.code.ClassFinder.fillIn(ClassFinder.java:373)
> 	at jdk.compiler/com.sun.tools.javac.code.ClassFinder.complete(ClassFinder.java:302)
> 	at jdk.compiler/com.sun.tools.javac.code.Symtab$2.complete(Symtab.java:360)
> 	at jdk.compiler/com.sun.tools.javac.code.Symbol.complete(Symbol.java:682)
> 	at jdk.compiler/com.sun.tools.javac.code.Symbol$ClassSymbol.complete(Symbol.java:1418)
> 	at jdk.compiler/com.sun.tools.javac.code.Symbol$ClassSymbol.flags(Symbol.java:1334)
> 	at jdk.compiler/com.sun.tools.javac.code.Types$12.visitClassType(Types.java:2183)
> 	at jdk.compiler/com.sun.tools.javac.code.Types$12.visitClassType(Types.java:2159)
> 	at jdk.compiler/com.sun.tools.javac.code.Type$ClassType.accept(Type.java:1050)
> 	at jdk.compiler/com.sun.tools.javac.code.Types$DefaultTypeVisitor.visit(Types.java:4894)
> 	at jdk.compiler/com.sun.tools.javac.code.Types.asSuper(Types.java:2156)
> 	at jdk.compiler/com.sun.tools.javac.code.Types$12.visitClassType(Types.java:2179)
> 	at jdk.compiler/com.sun.tools.javac.code.Types$12.visitClassType(Types.java:2159)
> 	at jdk.compiler/com.sun.tools.javac.code.Type$ClassType.accept(Type.java:1050)
> ```

This issue was the result of a change to include the a union inside `ResolvedMethodEntry`. The most recent commit and merge should have addressed this problem and made the interpreter code stable. Thanks for your patience

coleenp commented on src/hotspot/share/interpreter/rewriter.hpp:
```diff
@@ -68,24 +70,14 @@ class Rewriter: public StackObj {
-    assert(_first_iteration_cp_cache_limit != -1, "only valid after first iteration");
-    return _cp_cache_map.length() - _first_iteration_cp_cache_limit;
-  }
-
```
I'm so pleased this special case went away.  We still need to add entries for invokespecial of JVM_CONSTANT_InterfaceMethodref but we don't need to keep track of where the original entries ended anymore with your changes.

coleenp commented on src/hotspot/share/interpreter/templateTable.hpp:
```diff
@@ -262,20 +262,33 @@ class TemplateTable: AllStatic {
+  static void load_resolved_method_entry_interface(Register cache,
+                                               Register klass,
+                                               Register method_or_table_index,
+                                               Register flags);
```
Nit: can you line up these arguments?  You did move these. thank you.

coleenp commented on src/hotspot/share/oops/cpCache.cpp:
```diff
@@ -300,394 +219,165 @@ void ConstantPoolCacheEntry::set_direct_or_vtable_call(Bytecodes::Code invoke_co
 
-void ConstantPoolCacheEntry::set_itable_call(Bytecodes::Code invoke_code,
+void ConstantPoolCache::set_itable_call(Bytecodes::Code invoke_code,
+                                             int method_index,
```
Nit: align parameters here.

coleenp commented on src/hotspot/share/oops/resolvedMethodEntry.cpp:
```diff
@@ -0,0 +1,91 @@
+      assert(_method->is_valid() && _method->is_method(), "m is a valid method");
+      return !_method->is_old() && !_method->is_obsolete(); // old is always set for old and obsolete
+  } else {
+      return true;
```
nit: indent 2.

coleenp commented on src/hotspot/share/oops/resolvedMethodEntry.hpp:
```diff
@@ -0,0 +1,220 @@
+#define SHARE_OOPS_RESOLVEDMETHODENTRY_HPP
+
+#include "interpreter/bytecodes.hpp"
+#include "oops/instanceKlass.hpp"
```
Can you forward declare InstanceKlass rather than including it here?

coleenp reviewed (approved):
This is really great work and a wonderful cleanup of all these special cases we've had to hack into this over the years.  My comments are alignment nits, plus not including instanceKlass.hpp if you don't need to.  Really nice.

RealFYang commented on src/hotspot/cpu/aarch64/templateTable_aarch64.cpp:
```diff
@@ -3242,68 +3288,33 @@ void TemplateTable::fast_xaccess(TosState state)
   // compute return type
   // x86 uses a shift and mask or wings it with a shift plus assert
   // the mask is not needed. aarch64 just uses bitfield extract
-  __ ubfxw(rscratch2, flags, ConstantPoolCacheEntry::tos_state_shift,  ConstantPoolCacheEntry::tos_state_bits);
```
Nit: You might want to remove the preceding 3-line code comment together.

offamitkumar commented on src/hotspot/cpu/aarch64/templateTable_aarch64.cpp:
```diff
@@ -3242,68 +3288,29 @@ void TemplateTable::fast_xaccess(TosState state)
-                                   Register index,  // itable index, MethodType, etc.
-                                   Register recv,   // if caller wants to see it
+void TemplateTable::prepare_invoke(Register recv,   // if caller wants to see it
                                    Register flags   // if caller wants to test it
```
Hi All, 

I think `flags` argument here is unnecessary, as it is not being in this method. 

similarly `invokestatic` & `invokespecial` method is getting flags value from `load_resolved_method_entry_special_or_static` but it's not using it.

adinn reviewed (approved):
Thanks Matias. All my comments have been addressed.

matias9927 commented:
> Hi, @RealFYang and I have finished the RISC-V part, tier1-3 and hotspot:tier4 tested on hifive unmatched board. Please help us to add the RISC-V part, thanks a lot! [15455-riscv-port.diff.txt](https://github.com/openjdk/jdk/files/13214653/15455-riscv-port.diff.txt)

Thank you for the help!
/contributor add @zifeihan
/contributor add @RealFYang

TheRealMDoerr reviewed (commented):
I'll work on the PPC64 implementation. I have some minor suggestions regarding the aarch64 code after comparing it to x86.

matias9927 commented on src/hotspot/cpu/aarch64/templateTable_aarch64.cpp:
```diff
@@ -2357,32 +2366,107 @@ void TemplateTable::load_resolved_field_entry(Register obj,
-  __ ldrw(flags, Address(cache, in_bytes(cp_base_offset +
-                                           ConstantPoolCacheEntry::flags_offset())));
+  // setup registers
+  const Register index = r4;
```
This exists in x86 as well in each of the `load_resolved_method_entry_...()` methods. Some of these only have three arguments which cannot be reused so there is the option to include `index` as an argument, but this introduces an inconsistency among these similar methods. 

Should all of these methods take `index` which can be a reused register?

adinn replied:
If you add an extra argument then please name it `temp` or something equally clear as to the fact that this is for storing a local scratch register rather than an input/output. n.b. this is the main reason for resisting the urge to move register declarations+initializations up the call chain. It makes it less obvious that a register is only being used local to a callee.

TheRealMDoerr replied:
I think `index` is only used before `flags` and `method` get assigned. Can't we use one of them?
Otherwise, passing `temp` would be fine, too.

matias9927 commented on src/hotspot/cpu/aarch64/templateTable_aarch64.cpp:
```diff
@@ -3242,68 +3288,26 @@ void TemplateTable::fast_xaccess(TosState state)
-  // determine flags
+void TemplateTable::prepare_invoke(Register recv) {
+
+  const Register cache = r2;
```
As @offamitkumar pointed out, `flags` is unused in aarch64.

TheRealMDoerr replied:
Ok, then only `cache`.

zifeihan commented:
@matias9927  Some extra change to keep the riscv part up to date, please help us to add the RISC-V part.

[15455-riscv-port-v2.diff.txt](https://github.com/openjdk/jdk/files/13247611/15455-riscv-port-v2.diff.txt)

TheRealMDoerr commented:
I have a version which works for PPC64: https://github.com/TheRealMDoerr/jdk/commit/6bff39224e3129a898711a392b64c38b331d79a2

Note that I have implemented a few things slightly differently:

- `TemplateTable::load_resolved_method_entry_handle`: I'm loading the method at the end which avoids pushing and popping it on the expression stack which is not so nice IMHO. This works because I'm using a non-volatile register (asserted) for `cache` which is still valid after the C-call in `resolve_oop_handle`.
- `TemplateTable::load_resolved_method_entry_interface` and `TemplateTable::load_resolved_method_entry_virtual`: I'm not putting values in registers depending on the flags because it doesn't fit nicely into the PPC64 design. I found myself scratching my head and thinking about what is in the register in which case. Instead of that, I'm loading the fields where they are needed which leads to a much cleaner design. I always know what is in which register this way.

Please take a look and take these differences into consideration for other platforms. Thanks!

fparain reviewed (approved):
Thank you for this great rework. Looks good to me now.

matias9927 commented:
> I have a version which works for PPC64: [TheRealMDoerr@6bff392](https://github.com/TheRealMDoerr/jdk/commit/6bff39224e3129a898711a392b64c38b331d79a2)
> 
> Note that I have implemented a few things slightly differently:
> 
>     * `TemplateTable::load_resolved_method_entry_handle`: I'm loading the method at the end which avoids pushing and popping it on the expression stack which is not so nice IMHO. This works because I'm using a non-volatile register (asserted) for `cache` which is still valid after the C-call in `resolve_oop_handle`.
> 
>     * `TemplateTable::load_resolved_method_entry_interface` and `TemplateTable::load_resolved_method_entry_virtual`: I'm not putting values in registers depending on the flags because it doesn't fit nicely into the PPC64 design. I found myself scratching my head and thinking about what is in the register in which case. Instead of that, I'm loading the fields where they are needed which leads to a much cleaner design. I always know what is in which register this way.
> 
> 
> Please take a look and take these differences into consideration for other platforms. Thanks!

Thank you for the port! I liked your recommendation with regards to invokehandle and added that change to x86 and aarch64 as well.

/contributor add @TheRealMDoerr

zifeihan commented:
Hi @matias9927 : This change [1] for aarch64 and x86 looks nice and I am trying to prepare a similar one for riscv. But I have a small question regarding this change.
I see we are using `r2` for `cache` on aarch64 which is a volatile (or caller-saved) register according to the ABI. So the value in `cache` won't be preserved accross the C-call in resolve_oop_handle. Will this work? Is there anything I missed? Thanks.

[1] https://github.com/openjdk/jdk/pull/15455/commits/7d2a20ccf63f6140f7ba96904af2e3d89c6ab370

adinn commented:
@matias9927 aarch64 changes still look good after applying Martin's tweak.

TheRealMDoerr commented:
> Hi @matias9927 : This change [1] for aarch64 and x86 looks nice and I am trying to prepare a similar one for riscv. But I have a small question regarding this change. I see we are using `r2` for `cache` on aarch64 which is a volatile (or caller-saved) register according to the ABI. So the value in `cache` won't be preserved accross the C-call in resolve_oop_handle. Will this work? Is there anything I missed? Thanks.
> 
> [1] [7d2a20c](https://github.com/openjdk/jdk/commit/7d2a20ccf63f6140f7ba96904af2e3d89c6ab370)

Please note that I'm using `assert(cache->is_nonvolatile(), "C-call in resolve_oop_handle");` on PPC64 which makes it reliable.

zifeihan commented:
> > Hi @matias9927 : This change [1] for aarch64 and x86 looks nice and I am trying to prepare a similar one for riscv. But I have a small question regarding this change. I see we are using `r2` for `cache` on aarch64 which is a volatile (or caller-saved) register according to the ABI. So the value in `cache` won't be preserved accross the C-call in resolve_oop_handle. Will this work? Is there anything I missed? Thanks.
> > [1] [7d2a20c](https://github.com/openjdk/jdk/commit/7d2a20ccf63f6140f7ba96904af2e3d89c6ab370)
> 
> Please note that I'm using `assert(cache->is_nonvolatile(), "C-call in resolve_oop_handle");` on PPC64 which makes it reliable.

Hi Martin, Yes, I see that assertion for PPC. But my question is about aarch64 where `cache` could alias `r2` which is an volatile/caller-save register according to the aarch64 ABI. I am not sure if this is OK. Maybe I missed anything?

TheRealMDoerr commented:
> > > Hi @matias9927 : This change [1] for aarch64 and x86 looks nice and I am trying to prepare a similar one for riscv. But I have a small question regarding this change. I see we are using `r2` for `cache` on aarch64 which is a volatile (or caller-saved) register according to the ABI. So the value in `cache` won't be preserved accross the C-call in resolve_oop_handle. Will this work? Is there anything I missed? Thanks.
> > > [1] [7d2a20c](https://github.com/openjdk/jdk/commit/7d2a20ccf63f6140f7ba96904af2e3d89c6ab370)
> > 
> > 
> > Please note that I'm using `assert(cache->is_nonvolatile(), "C-call in resolve_oop_handle");` on PPC64 which makes it reliable.
> 
> Hi Martin, Yes, I see that assertion for PPC. But my question is about aarch64 where `cache` could alias `r2` which is an volatile/caller-save register according to the aarch64 ABI. I am not sure if this is OK. Maybe I missed anything?

Sorry, I was not clear enough. I wanted to suggest adding the assertion to other platforms and changing the register if needed.

TheRealMDoerr commented:
AFAICS, `resolve_oop_handle` preserves volatile regs on aarch64 and x86_64, so, the current version should be ok.
I can see the following test passing:
`make run-test TEST="java/lang/invoke" JTREG="VM_OPTIONS=-XX:+UseZGC -XX:+ZGenerational"` on linux x86_64, aarch64 and ppc64le.

zifeihan commented:
> AFAICS, `resolve_oop_handle` preserves volatile regs on aarch64 and x86_64, so, the current version should be ok. I can see the following test passing: `make run-test TEST="java/lang/invoke" JTREG="VM_OPTIONS=-XX:+UseZGC -XX:+ZGenerational"` on linux x86_64, aarch64 and ppc64le.

Thank you for taking a look. I also checked the aarch64 and x86 code and now I see the difference.

zifeihan commented:
@matias9927 Hi, I have prepared a similar improvement for riscv64. The volatile registers are also preserved on this platform. 
[15455-riscv-port-v3.diff.txt](https://github.com/openjdk/jdk/files/13319170/15455-riscv-port-v3.diff.txt)

TheRealMDoerr commented on src/hotspot/share/oops/resolvedMethodEntry.hpp:
```diff
@@ -0,0 +1,220 @@
+  u2 _number_of_parameters;          // Number of arguments for method
+  u1 _tos_state;                     // TOS state
+  u1 _flags;                         // Flags: [00|has_resolved_ref_index|has_local_signature|has_appendix|forced_virtual|final|virtual_final]
+  u1 _bytecode1, _bytecode2;         // Resovled invoke codes
```
Typo: Resovled

offamitkumar commented:
Hi @matias9927, Please add s390-Port from here: https://github.com/offamitkumar/jdk/commit/3f7017467b1a4ae8fe70530c7183c2667cf2c7f2

matias9927 commented:
> Hi @matias9927, Please add s390-Port from here: [offamitkumar@3f70174](https://github.com/offamitkumar/jdk/commit/3f7017467b1a4ae8fe70530c7183c2667cf2c7f2)

Thank you for the port Amit!

/contributor add @offamitkumar

coleenp reviewed (approved):
Does it need another approval.  If so, here it is.

vnkozlov commented:
@matias9927 this broke arm (32-bit) build:
```
/home/runner/work/jdk/jdk/src/hotspot/cpu/arm/interp_masm_arm.cpp:236:17: error: 'ConstantPoolCacheEntry' was not declared in this scope; did you mean 'ConstantPoolCache'?
  236 |   assert(sizeof(ConstantPoolCacheEntry) == 4*wordSize, "adjust code below");
      |                 ^~~~~~~~~~~~~~~~~~~~~~
```
I don't see GHA testing is setup for your repo.

matias9927 commented:
> @matias9927 this broke arm (32-bit) build:
> 
> ```
> /home/runner/work/jdk/jdk/src/hotspot/cpu/arm/interp_masm_arm.cpp:236:17: error: 'ConstantPoolCacheEntry' was not declared in this scope; did you mean 'ConstantPoolCache'?
>   236 |   assert(sizeof(ConstantPoolCacheEntry) == 4*wordSize, "adjust code below");
>       |                 ^~~~~~~~~~~~~~~~~~~~~~
> ```
> 
> I don't see GHA testing is setup for your repo.

This patch includes changes to the interpreters which were provided by porters. Oracle is responsible for x86-64 and aarch64 so completed those and made an effort to inform porters of the upcoming change. The ARM32 port has not yet been completed and thus is not part of this patch.

The ARM32 ports have previously been handled by @bulasevich and @voitylov and they have been contacted.

voitylov commented:
Noted, we'll follow up with the arm32 fix a little later. Thanks Matias!

matias9927 commented:
> Noted, we'll follow up with the arm32 fix a little later. Thanks Matias!

Thanks for the confirmation @voitylov, I look forward to the port!

tstuefe commented:
@matias9927 Having the arm *build* broken is really bad. It's one thing if the VM is dead on arrival (bad enough) but this shows up in everyone's GHA as a red flag. It teaches people to ignore GHAs. We require clean GHAs from outside developers; I think this sets a bad precedence.

andrew-m-leonard commented:
Adoptium also unable to build jdk-22 ARM32, I have raised a bug: https://bugs.openjdk.org/browse/JDK-8320402

sunny868 commented on src/hotspot/cpu/riscv/templateTable_riscv.cpp:
```diff
@@ -2168,21 +2169,32 @@ void TemplateTable::_return(TosState state) {
+      break;
+  }
+  // Load-acquire the bytecode to match store-release in InterpreterRuntime
+  __ membar(MacroAssembler::AnyAny);
```
Perhaps just using `StoreLoad` here is enough, right?
